File: js/src/examples/quick-start.ts
/**
 * Quick Start Example for TOPAY-Z512
 * 
 * This example demonstrates the basic usage of TOPAY-Z512 cryptographic library
 * including key generation, hashing, KEM operations, and data fragmentation.
 */

import {
  generateKeyPair,
  computeHash,
  kemKeyGen,
  kemEncapsulate,
  kemDecapsulate,
  fragmentData,
  reconstructData,
  toHex
} from '../index';

async function quickStartExample(): Promise<void> {
  console.log('🚀 TOPAY-Z512 Quick Start Example');
  console.log('='.repeat(50));

  try {
    // 1. Generate a key pair
    console.log('\n1. Generating key pair...');
    const keyPair = await generateKeyPair();
    console.log(`   Private key: ${toHex(keyPair.privateKey).substring(0, 32)}...`);
    console.log(`   Public key:  ${toHex(keyPair.publicKey).substring(0, 32)}...`);

    // 2. Hash some data
    console.log('\n2. Computing hash...');
    const testData = new TextEncoder().encode('Hello, TOPAY-Z512!');
    const hash = await computeHash(testData);
    console.log(`   Data: "Hello, TOPAY-Z512!"`);
    console.log(`   Hash: ${toHex(hash).substring(0, 32)}...`);

    // 3. KEM operations
    console.log('\n3. KEM operations...');
    const kemKeyPair = await kemKeyGen();
    console.log('   Generated KEM key pair');

    const { ciphertext, sharedSecret } = await kemEncapsulate(kemKeyPair.publicKey);
    console.log('   Encapsulated shared secret');
    console.log(`   Ciphertext: ${toHex(ciphertext).substring(0, 32)}...`);
    console.log(`   Shared secret: ${toHex(sharedSecret).substring(0, 32)}...`);

    const decapsulatedSecret = await kemDecapsulate(kemKeyPair.secretKey, ciphertext);
    console.log('   Decapsulated shared secret');
    console.log(`   Secrets match: ${toHex(sharedSecret) === toHex(decapsulatedSecret)}`);

    // 4. Data fragmentation
    console.log('\n4. Data fragmentation...');
    const largeData = new Uint8Array(1024);
    crypto.getRandomValues(largeData);
    
    const fragResult = await fragmentData(largeData);
    console.log(`   Original size: ${fragResult.metadata.originalSize} bytes`);
    console.log(`   Fragment count: ${fragResult.metadata.fragmentCount}`);
    console.log(`   Checksum: ${toHex(fragResult.metadata.checksum).substring(0, 32)}...`);

    const reconResult = await reconstructData(fragResult.fragments);
    console.log(`   Reconstruction complete: ${reconResult.isComplete}`);
    console.log(`   Data integrity: ${largeData.length === reconResult.data.length}`);

    console.log('\n✅ Quick start example completed successfully!');
    console.log('\nNext steps:');
    console.log('- Explore other examples in the examples/ directory');
    console.log('- Read the documentation for advanced features');
    console.log('- Check out the benchmark example for performance testing');

  } catch (error) {
    console.error('❌ Error in quick start example:', error);
    process.exit(1);
  }
}

// Run the example
if (require.main === module) {
  quickStartExample().catch(console.error);
}

export { quickStartExample };

File: rust/examples/quick_start.rs
// Quick Start Example for TOPAY-Z512
// This shows the most common use cases in a simple format

use topayz512::{Hash, Kem, KeyPair};

fn main() {
    println!("🚀 TOPAY-Z512 Quick Start Guide");
    println!("================================\n");

    // 1. Generate a key pair
    println!("1. Key Pair Generation:");
    let keypair = KeyPair::generate();
    println!("   ✅ Generated new key pair");
    println!(
        "   Private key (first 16 chars): {}",
        &keypair.private_key().to_hex()[..16]
    );
    println!(
        "   Public key (first 16 chars):  {}\n",
        &keypair.public_key().to_hex()[..16]
    );

    // 2. Hash some data
    println!("2. Hashing:");
    let message = b"Hello, TOPAY-Z512!";
    let hash = Hash::new(message);
    println!("   Message: {:?}", std::str::from_utf8(message).unwrap());
    println!("   Hash (first 32 chars): {}\n", &hash.to_hex()[..32]);

    // 3. Key Encapsulation Mechanism
    println!("3. Key Encapsulation:");
    let (kem_public, kem_secret) = Kem::keygen();
    let (ciphertext, shared_secret1) = Kem::encapsulate(&kem_public);
    let shared_secret2 = Kem::decapsulate(&kem_secret, &ciphertext);

    println!("   ✅ Generated KEM keys");
    println!("   ✅ Encapsulated shared secret");
    println!("   ✅ Decapsulated shared secret");
    println!(
        "   Secrets match: {}\n",
        shared_secret1.to_bytes() == shared_secret2.to_bytes()
    );

    // 4. Serialization
    println!("4. Serialization:");
    let (private_hex, public_hex) = keypair.to_hex();
    println!("   ✅ Serialized keys to hex");
    println!("   Private hex length: {} chars", private_hex.len());
    println!("   Public hex length:  {} chars\n", public_hex.len());

    // 5. Deserialization
    println!("5. Deserialization:");
    let restored_private = topayz512::keypair::PrivateKey::from_hex(&private_hex).unwrap();
    let restored_public = topayz512::keypair::PublicKey::from_hex(&public_hex).unwrap();
    println!("   ✅ Restored keys from hex");
    println!(
        "   Keys match: {}\n",
        *keypair.private_key() == restored_private && *keypair.public_key() == restored_public
    );

    println!("🎉 Quick start complete! You're ready to use TOPAY-Z512.");
    println!("\nNext steps:");
    println!(
        "- Run other examples: cargo run --example [keypair_example|kem_example|hash_example]"
    );
    println!("- Run tests: cargo test");
    println!("- Build for production: cargo build --release");
}
